### DISCUSSION TITLE
External Processes with cFS

### DISCUSSION THREAD

**Discussion by tstern-masten** [nasa/cFS#444] [Q&A] [answered]
Hello, 

I am a little confused about the way to interface cFS with external processes

The examples I have seen thus far are always with shared objects being created and started by cFS but what would be the mechanism to interface an entirely separate process with cFS? 

I have been reading up on External Code Interface but that seems to be targeting auto-generated or legacy code sources and not necessarily custom written applications. 

The hope is we would be able to run an application completely independent of cFS so if a fault were to happen that process would be unaffected, but could still utilize cFS services.

Thank you,
Tim

**Comment by skliper** [comment, marked as answer]
I see three options for interfacing processes based on what you need in terms of "utilizing cFS services".

Wrap both processes in the cFS framework, communicate through SBN.  Then each process has the full cFS set of core capabilities.

One cFS instance (with SBN), the other process just has an SBN compatible interface.  I thought there was an example of this somewhere or a related repo but I haven't found it yet... maybe someone in the community could provide a pointer?

Or make a custom app to run on the full cFS instance that interfaces with the other process through whatever custom mechanism fits your use case.  For example if the custom process produces/consumes a data stream make an app that talks the same language and converts to CCSDS messages on the cFS side (if that's what you want).  Could be message queues, DMA, shared memory, etc. whatever works for that custom process.

**Reply by tstern-masten** [reply]
Thank you for the response. 

The third option sounds very similar to what ECI might be able to do? I might explore either integrating that or creating a custom interface. 

The second option is interesting as well. If you do find that example I would be interested in how that looks. 

Again thank you for the help!

**Reply by skliper** [reply]
If you want an interface less complex than SBN, you could just use something like ci_lab/to_lab to communicate over a stream.  Although just a simple point to point message queue might also be easy.  You just need something simple on the cFS process to move the data onto/off of the software bus (if that's what your use case is).

**Reply by rlourette** [reply]
I believe that is called SBN-Client and it is in github

**Comment by lark1318** [comment]
You might want to look at https://github.com/nasa/sbn-client
I think it was created for the case you mentioned

**Reply by tstern-masten** [reply]
Thank you! I thought I remembered something like this but was having trouble finding it again. 

Do you happen to know what would preclude someone from using this in flight. I see the note in the readme that it is "not intended to be used in flight software" and I am just wondering why that might be. 

Thank you again!
Tim
